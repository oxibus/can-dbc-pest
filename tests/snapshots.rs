#![cfg(feature = "encoding")]

use std::borrow::Cow;
use std::fs;
use std::path::Path;

use can_dbc_pest::{decode_cp1252, DbcParser, Rule};
use insta::{assert_debug_snapshot, with_settings};
use pest::Parser;

/// Test parsing all DBC files
#[test]
fn test_cantools_dbc_files() {
    let test_dirs = [
        (
            "tests/shared-test-files/dbc-cantools",
            "dbc-cantools",
            true,
            true,
        ),
        ("tests/shared-test-files/canpy", "canpy", true, true),
        ("tests/opendbc/opendbc/dbc", "opendbc", false, false),
    ];
    for (path, file_name, snapshot, cp1251) in test_dirs {
        let path = Path::new(path);
        let snapshot_path = String::from("snapshots-") + file_name;
        // snapshots should go to /tests/snapshots-* directory
        with_settings! {
            { omit_expression => true,
              snapshot_path => snapshot_path ,
              prepend_module_to_snapshot => false },
            {
                test_dbc_files(path, snapshot, cp1251);
            }
        }
    }
}

/// Test parsing all DBC files in the given directory.
fn test_dbc_files(dir: impl AsRef<Path>, snapshot: bool, cp1251: bool) {
    let dir = dir.as_ref();
    let dir_display = dir.display();
    let dir_content = fs::read_dir(dir).unwrap_or_else(|e| {
        panic!(
            "
--------------------------------------------------------------------------
Error reading dbc test files from   {dir_display}
{e}
Make sure git submodules are up to date by running
    git submodule update --init --recursive
--------------------------------------------------------------------------
"
        )
    });

    eprintln!("Testing dbc files in directory: {dir_display}");
    for dbc_path in dir_content {
        let path = dbc_path.unwrap().path();
        if path
            .extension()
            .is_some_and(|ext| ext.eq_ignore_ascii_case("dbc"))
        {
            parse_one_file(path, snapshot, cp1251);
        }
    }
}

/// Parse a single DBC file and assert a snapshot of the result.
fn parse_one_file(path: impl AsRef<Path>, snapshot: bool, cp1251: bool) {
    let path = path.as_ref();
    eprintln!("Testing DBC file: {}", path.display());
    let file_name = path.file_stem().unwrap().to_string_lossy().to_string();
    let buffer = fs::read(path).unwrap();
    let buffer = if cp1251 {
        decode_cp1252(&buffer)
            .unwrap_or_else(|| panic!("Failed to decode {} as cp1252", path.display()))
    } else {
        Cow::Borrowed(
            std::str::from_utf8(&buffer)
                .unwrap_or_else(|_| panic!("Failed to decode {} as utf-8", path.display())),
        )
    };

    match DbcParser::parse(Rule::file, &buffer) {
        Ok(pairs) => {
            if std::env::var("SKIP_INSTA").is_err()
                && (snapshot || std::env::var("FORCE_INSTA").is_ok())
            {
                assert_debug_snapshot!(file_name, pairs);
            }
        }
        Err(e) => panic!("Failed to parse {file_name}.dbc: {e:#?}"),
    }
}
