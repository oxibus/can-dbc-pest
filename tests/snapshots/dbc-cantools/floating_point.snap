---
source: tests/snapshots.rs
---
[
    Pair {
        rule: file,
        span: Span {
            str: "VERSION \"HIPBNYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYY/4/%%%/4/'%**4YYY///\"\n\n\nNS_ : \n\tNS_DESC_\n\tCM_\n\tBA_DEF_\n\tBA_\n\tVAL_\n\tCAT_DEF_\n\tCAT_\n\tFILTER\n\tBA_DEF_DEF_\n\tEV_DATA_\n\tENVVAR_DATA_\n\tSGTYPE_\n\tSGTYPE_VAL_\n\tBA_DEF_SGTYPE_\n\tBA_SGTYPE_\n\tSIG_TYPE_REF_\n\tVAL_TABLE_\n\tSIG_GROUP_\n\tSIG_VALTYPE_\n\tSIGTYPE_VALTYPE_\n\tBO_TX_BU_\n\tBA_DEF_REL_\n\tBA_REL_\n\tBA_DEF_DEF_REL_\n\tBU_SG_REL_\n\tBU_EV_REL_\n\tBU_BO_REL_\n\tSG_MUL_VAL_\n\nBS_:\n\nBU_: TestNode\n\n\nBO_ 1024 Message1: 8 TestNode\n SG_ Signal1 : 0|64@1- (1,0) [0|0] \"\" Vector__XXX\n\nBO_ 1025 Message2: 8 TestNode\n SG_ Signal2 : 32|32@1- (1,0) [0|0] \"\" Vector__XXX\n SG_ Signal1 : 0|32@1- (1,0) [0|0] \"\" Vector__XXX\n\n\n\n\nCM_ BU_ TestNode \"\";\nBA_DEF_ BO_  \"GenMsgCycleTime\" INT 0 65535;\nBA_DEF_DEF_  \"GenMsgCycleTime\" 0;\n\n\nSIG_VALTYPE_ 1024 Signal1 : 2;\nSIG_VALTYPE_ 1025 Signal1 : 1;\nSIG_VALTYPE_ 1025 Signal2 : 1;\n\n\n",
            start: 0,
            end: 833,
        },
        inner: [
            Pair {
                rule: version,
                span: Span {
                    str: "VERSION \"HIPBNYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYY/4/%%%/4/'%**4YYY///\"\n",
                    start: 0,
                    end: 73,
                },
                inner: [
                    Pair {
                        rule: quoted_str,
                        span: Span {
                            str: "\"HIPBNYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYY/4/%%%/4/'%**4YYY///\"",
                            start: 8,
                            end: 72,
                        },
                        inner: [
                            Pair {
                                rule: string,
                                span: Span {
                                    str: "HIPBNYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYY/4/%%%/4/'%**4YYY///",
                                    start: 9,
                                    end: 71,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: new_symbols,
                span: Span {
                    str: "NS_ : \n\tNS_DESC_\n\tCM_\n\tBA_DEF_\n\tBA_\n\tVAL_\n\tCAT_DEF_\n\tCAT_\n\tFILTER\n\tBA_DEF_DEF_\n\tEV_DATA_\n\tENVVAR_DATA_\n\tSGTYPE_\n\tSGTYPE_VAL_\n\tBA_DEF_SGTYPE_\n\tBA_SGTYPE_\n\tSIG_TYPE_REF_\n\tVAL_TABLE_\n\tSIG_GROUP_\n\tSIG_VALTYPE_\n\tSIGTYPE_VALTYPE_\n\tBO_TX_BU_\n\tBA_DEF_REL_\n\tBA_REL_\n\tBA_DEF_DEF_REL_\n\tBU_SG_REL_\n\tBU_EV_REL_\n\tBU_BO_REL_\n\tSG_MUL_VAL_\n\n",
                    start: 75,
                    end: 399,
                },
                inner: [
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "NS_DESC_",
                            start: 83,
                            end: 91,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "CM_",
                            start: 93,
                            end: 96,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_DEF_",
                            start: 98,
                            end: 105,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_",
                            start: 107,
                            end: 110,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "VAL_",
                            start: 112,
                            end: 116,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "CAT_DEF_",
                            start: 118,
                            end: 126,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "CAT_",
                            start: 128,
                            end: 132,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "FILTER",
                            start: 134,
                            end: 140,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_DEF_DEF_",
                            start: 142,
                            end: 153,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "EV_DATA_",
                            start: 155,
                            end: 163,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "ENVVAR_DATA_",
                            start: 165,
                            end: 177,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SGTYPE_",
                            start: 179,
                            end: 186,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SGTYPE_VAL_",
                            start: 188,
                            end: 199,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_DEF_SGTYPE_",
                            start: 201,
                            end: 215,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_SGTYPE_",
                            start: 217,
                            end: 227,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SIG_TYPE_REF_",
                            start: 229,
                            end: 242,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "VAL_TABLE_",
                            start: 244,
                            end: 254,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SIG_GROUP_",
                            start: 256,
                            end: 266,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SIG_VALTYPE_",
                            start: 268,
                            end: 280,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SIGTYPE_VALTYPE_",
                            start: 282,
                            end: 298,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BO_TX_BU_",
                            start: 300,
                            end: 309,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_DEF_REL_",
                            start: 311,
                            end: 322,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_REL_",
                            start: 324,
                            end: 331,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BA_DEF_DEF_REL_",
                            start: 333,
                            end: 348,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BU_SG_REL_",
                            start: 350,
                            end: 360,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BU_EV_REL_",
                            start: 362,
                            end: 372,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "BU_BO_REL_",
                            start: 374,
                            end: 384,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: ident,
                        span: Span {
                            str: "SG_MUL_VAL_",
                            start: 386,
                            end: 397,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: bit_timing,
                span: Span {
                    str: "BS_:\n",
                    start: 399,
                    end: 404,
                },
                inner: [],
            },
            Pair {
                rule: nodes,
                span: Span {
                    str: "BU_: TestNode\n\n\n",
                    start: 405,
                    end: 421,
                },
                inner: [
                    Pair {
                        rule: node_name,
                        span: Span {
                            str: "TestNode",
                            start: 410,
                            end: 418,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: message,
                span: Span {
                    str: "BO_ 1024 Message1: 8 TestNode",
                    start: 421,
                    end: 450,
                },
                inner: [
                    Pair {
                        rule: msg_var,
                        span: Span {
                            str: "BO_ 1024",
                            start: 421,
                            end: 429,
                        },
                        inner: [
                            Pair {
                                rule: message_id,
                                span: Span {
                                    str: "1024",
                                    start: 425,
                                    end: 429,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: message_name,
                        span: Span {
                            str: "Message1",
                            start: 430,
                            end: 438,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: message_size,
                        span: Span {
                            str: "8",
                            start: 440,
                            end: 441,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: transmitter,
                        span: Span {
                            str: "TestNode",
                            start: 442,
                            end: 450,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: signal,
                span: Span {
                    str: "SG_ Signal1 : 0|64@1- (1,0) [0|0] \"\" Vector__XXX",
                    start: 452,
                    end: 500,
                },
                inner: [
                    Pair {
                        rule: signal_name,
                        span: Span {
                            str: "Signal1",
                            start: 456,
                            end: 463,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: start_bit,
                        span: Span {
                            str: "0",
                            start: 466,
                            end: 467,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: signal_size,
                        span: Span {
                            str: "64",
                            start: 468,
                            end: 470,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: little_endian,
                        span: Span {
                            str: "1",
                            start: 471,
                            end: 472,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: signed_type,
                        span: Span {
                            str: "-",
                            start: 472,
                            end: 473,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: factor,
                        span: Span {
                            str: "1",
                            start: 475,
                            end: 476,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: offset,
                        span: Span {
                            str: "0",
                            start: 477,
                            end: 478,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: min_max,
                        span: Span {
                            str: "[0|0]",
                            start: 480,
                            end: 485,
                        },
                        inner: [
                            Pair {
                                rule: minimum,
                                span: Span {
                                    str: "0",
                                    start: 481,
                                    end: 482,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: maximum,
                                span: Span {
                                    str: "0",
                                    start: 483,
                                    end: 484,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: unit,
                        span: Span {
                            str: "\"\"",
                            start: 486,
                            end: 488,
                        },
                        inner: [
                            Pair {
                                rule: string,
                                span: Span {
                                    str: "",
                                    start: 487,
                                    end: 487,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: node_name,
                        span: Span {
                            str: "Vector__XXX",
                            start: 489,
                            end: 500,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: message,
                span: Span {
                    str: "BO_ 1025 Message2: 8 TestNode",
                    start: 502,
                    end: 531,
                },
                inner: [
                    Pair {
                        rule: msg_var,
                        span: Span {
                            str: "BO_ 1025",
                            start: 502,
                            end: 510,
                        },
                        inner: [
                            Pair {
                                rule: message_id,
                                span: Span {
                                    str: "1025",
                                    start: 506,
                                    end: 510,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: message_name,
                        span: Span {
                            str: "Message2",
                            start: 511,
                            end: 519,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: message_size,
                        span: Span {
                            str: "8",
                            start: 521,
                            end: 522,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: transmitter,
                        span: Span {
                            str: "TestNode",
                            start: 523,
                            end: 531,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: signal,
                span: Span {
                    str: "SG_ Signal2 : 32|32@1- (1,0) [0|0] \"\" Vector__XXX",
                    start: 533,
                    end: 582,
                },
                inner: [
                    Pair {
                        rule: signal_name,
                        span: Span {
                            str: "Signal2",
                            start: 537,
                            end: 544,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: start_bit,
                        span: Span {
                            str: "32",
                            start: 547,
                            end: 549,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: signal_size,
                        span: Span {
                            str: "32",
                            start: 550,
                            end: 552,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: little_endian,
                        span: Span {
                            str: "1",
                            start: 553,
                            end: 554,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: signed_type,
                        span: Span {
                            str: "-",
                            start: 554,
                            end: 555,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: factor,
                        span: Span {
                            str: "1",
                            start: 557,
                            end: 558,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: offset,
                        span: Span {
                            str: "0",
                            start: 559,
                            end: 560,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: min_max,
                        span: Span {
                            str: "[0|0]",
                            start: 562,
                            end: 567,
                        },
                        inner: [
                            Pair {
                                rule: minimum,
                                span: Span {
                                    str: "0",
                                    start: 563,
                                    end: 564,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: maximum,
                                span: Span {
                                    str: "0",
                                    start: 565,
                                    end: 566,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: unit,
                        span: Span {
                            str: "\"\"",
                            start: 568,
                            end: 570,
                        },
                        inner: [
                            Pair {
                                rule: string,
                                span: Span {
                                    str: "",
                                    start: 569,
                                    end: 569,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: node_name,
                        span: Span {
                            str: "Vector__XXX",
                            start: 571,
                            end: 582,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: signal,
                span: Span {
                    str: "SG_ Signal1 : 0|32@1- (1,0) [0|0] \"\" Vector__XXX",
                    start: 584,
                    end: 632,
                },
                inner: [
                    Pair {
                        rule: signal_name,
                        span: Span {
                            str: "Signal1",
                            start: 588,
                            end: 595,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: start_bit,
                        span: Span {
                            str: "0",
                            start: 598,
                            end: 599,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: signal_size,
                        span: Span {
                            str: "32",
                            start: 600,
                            end: 602,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: little_endian,
                        span: Span {
                            str: "1",
                            start: 603,
                            end: 604,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: signed_type,
                        span: Span {
                            str: "-",
                            start: 604,
                            end: 605,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: factor,
                        span: Span {
                            str: "1",
                            start: 607,
                            end: 608,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: offset,
                        span: Span {
                            str: "0",
                            start: 609,
                            end: 610,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: min_max,
                        span: Span {
                            str: "[0|0]",
                            start: 612,
                            end: 617,
                        },
                        inner: [
                            Pair {
                                rule: minimum,
                                span: Span {
                                    str: "0",
                                    start: 613,
                                    end: 614,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: maximum,
                                span: Span {
                                    str: "0",
                                    start: 615,
                                    end: 616,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: unit,
                        span: Span {
                            str: "\"\"",
                            start: 618,
                            end: 620,
                        },
                        inner: [
                            Pair {
                                rule: string,
                                span: Span {
                                    str: "",
                                    start: 619,
                                    end: 619,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: node_name,
                        span: Span {
                            str: "Vector__XXX",
                            start: 621,
                            end: 632,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: comment,
                span: Span {
                    str: "CM_ BU_ TestNode \"\";",
                    start: 637,
                    end: 657,
                },
                inner: [
                    Pair {
                        rule: node_var,
                        span: Span {
                            str: "BU_ TestNode",
                            start: 641,
                            end: 653,
                        },
                        inner: [
                            Pair {
                                rule: node_name,
                                span: Span {
                                    str: "TestNode",
                                    start: 645,
                                    end: 653,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: quoted_str,
                        span: Span {
                            str: "\"\"",
                            start: 654,
                            end: 656,
                        },
                        inner: [
                            Pair {
                                rule: string,
                                span: Span {
                                    str: "",
                                    start: 655,
                                    end: 655,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: attr_def,
                span: Span {
                    str: "BA_DEF_ BO_  \"GenMsgCycleTime\" INT 0 65535;",
                    start: 658,
                    end: 701,
                },
                inner: [
                    Pair {
                        rule: object_type,
                        span: Span {
                            str: "BO_",
                            start: 666,
                            end: 669,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: attribute_name,
                        span: Span {
                            str: "\"GenMsgCycleTime\"",
                            start: 671,
                            end: 688,
                        },
                        inner: [
                            Pair {
                                rule: string,
                                span: Span {
                                    str: "GenMsgCycleTime",
                                    start: 672,
                                    end: 687,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: attribute_type_int,
                        span: Span {
                            str: "INT 0 65535",
                            start: 689,
                            end: 700,
                        },
                        inner: [
                            Pair {
                                rule: minimum,
                                span: Span {
                                    str: "0",
                                    start: 693,
                                    end: 694,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: maximum,
                                span: Span {
                                    str: "65535",
                                    start: 695,
                                    end: 700,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: ba_def_def,
                span: Span {
                    str: "BA_DEF_DEF_  \"GenMsgCycleTime\" 0;",
                    start: 702,
                    end: 735,
                },
                inner: [
                    Pair {
                        rule: attribute_name,
                        span: Span {
                            str: "\"GenMsgCycleTime\"",
                            start: 715,
                            end: 732,
                        },
                        inner: [
                            Pair {
                                rule: string,
                                span: Span {
                                    str: "GenMsgCycleTime",
                                    start: 716,
                                    end: 731,
                                },
                                inner: [],
                            },
                        ],
                    },
                    Pair {
                        rule: num_str_value,
                        span: Span {
                            str: "0",
                            start: 733,
                            end: 734,
                        },
                        inner: [
                            Pair {
                                rule: number,
                                span: Span {
                                    str: "0",
                                    start: 733,
                                    end: 734,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: signal_value_type,
                span: Span {
                    str: "SIG_VALTYPE_ 1024 Signal1 : 2;",
                    start: 738,
                    end: 768,
                },
                inner: [
                    Pair {
                        rule: message_id,
                        span: Span {
                            str: "1024",
                            start: 751,
                            end: 755,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: signal_name,
                        span: Span {
                            str: "Signal1",
                            start: 756,
                            end: 763,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "2",
                            start: 766,
                            end: 767,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: signal_value_type,
                span: Span {
                    str: "SIG_VALTYPE_ 1025 Signal1 : 1;",
                    start: 769,
                    end: 799,
                },
                inner: [
                    Pair {
                        rule: message_id,
                        span: Span {
                            str: "1025",
                            start: 782,
                            end: 786,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: signal_name,
                        span: Span {
                            str: "Signal1",
                            start: 787,
                            end: 794,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "1",
                            start: 797,
                            end: 798,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: signal_value_type,
                span: Span {
                    str: "SIG_VALTYPE_ 1025 Signal2 : 1;",
                    start: 800,
                    end: 830,
                },
                inner: [
                    Pair {
                        rule: message_id,
                        span: Span {
                            str: "1025",
                            start: 813,
                            end: 817,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: signal_name,
                        span: Span {
                            str: "Signal2",
                            start: 818,
                            end: 825,
                        },
                        inner: [],
                    },
                    Pair {
                        rule: int,
                        span: Span {
                            str: "1",
                            start: 828,
                            end: 829,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: EOI,
                span: Span {
                    str: "",
                    start: 833,
                    end: 833,
                },
                inner: [],
            },
        ],
    },
]
